From 7d5ea874dacb69dcd1a14e804d4271d2f776616a Mon Sep 17 00:00:00 2001
From: "GONG, Ruiqi" <gongruiqi1@huawei.com>
Date: Fri, 21 Jan 2022 17:28:40 +0800
Subject: [PATCH 1494/2065] [Huawei] selinux: genfscon ns: Retrieve isec
 according to process's namespace

Offering: HULK
hulk inclusion
category: feature
bugzilla: 186697

----------------------------------------

Currently it's been able to have multiple isec for an inode, and
processes can switch namespace already. So make processes query the
i_security list and retrieve the isec that corresponds to their own
namespace, or create one if no matching isec is found.

Signed-off-by: GONG, Ruiqi <gongruiqi1@huawei.com>
---
 security/selinux/hooks.c | 87 ++++++++++++++++++++++++++++++----------
 1 file changed, 65 insertions(+), 22 deletions(-)

diff --git a/security/selinux/hooks.c b/security/selinux/hooks.c
index ccd6d3c1b5bc..df059d576531 100644
--- a/security/selinux/hooks.c
+++ b/security/selinux/hooks.c
@@ -214,10 +214,13 @@ static bool inode_support_ns(const struct inode *inode)
 	return sbsec->flags & SE_SBSUPPORT_NS;
 }
 
-static struct inode_security_struct *isec_alloc(struct inode *inode)
+static struct inode_security_struct *isec_alloc(struct inode *inode,
+						bool may_sleep)
 {
-	struct inode_security_struct *isec = kmem_cache_zalloc(
-			inode_security_cache, GFP_NOFS);
+	struct inode_security_struct *isec;
+	gfp_t flags = may_sleep ? GFP_NOFS : GFP_NOWAIT;
+
+	isec = kmem_cache_zalloc(inode_security_cache, flags);
 	if (!isec)
 		return NULL;
 
@@ -340,10 +343,56 @@ static inline u32 task_sid(const struct task_struct *task)
 
 static int inode_doinit_with_dentry(struct inode *inode, struct dentry *opt_dentry);
 
+static struct inode_security_struct *find_isec_ns(struct inode *inode,
+						  bool may_sleep)
+{
+	struct inode_security_head *isech = selinux_inode(inode);
+	struct inode_security_struct *cur, *new;
+
+	if (!inode_support_ns(inode))
+		return container_of(isech->head.next,
+				    struct inode_security_struct,
+				    isec_list);
+
+	spin_lock(&isech->lock);
+
+	list_for_each_entry(cur, &isech->head, isec_list) {
+		if (cur->ns == current_selinux_ns)
+			goto unlock;
+	}
+
+	/* to avoid holding the lock when sleeping in isec_alloc */
+	spin_unlock(&isech->lock);
+
+	new = isec_alloc(inode, may_sleep);
+	if (!new) {
+		cur = NULL;
+		goto out;
+	}
+	/* task_sid on the whole chain should be the same */
+	new->task_sid = isech->task_sid;
+
+	spin_lock(&isech->lock);
+
+	list_for_each_entry(cur, &isech->head, isec_list) {
+		if (cur->ns == current_selinux_ns) {
+			kmem_cache_free(inode_security_cache, new);
+			goto unlock;
+		}
+	}
+
+	list_add(&new->isec_list, &isech->head);
+	cur = new;
+
+unlock:
+	spin_unlock(&isech->lock);
+out:
+	return cur;
+}
+
 struct inode_security_struct *inode_security_novalidate(struct inode *inode)
 {
-	return container_of(selinux_inode(inode)->head.next,
-			struct inode_security_struct, isec_list);
+	return find_isec_ns(inode, false);
 }
 
 /*
@@ -352,18 +401,19 @@ struct inode_security_struct *inode_security_novalidate(struct inode *inode)
  * allowed; when set to false, returns -ECHILD when the label is
  * invalid.  The @dentry parameter should be set to a dentry of the inode.
  */
-static int __inode_security_revalidate(struct inode *inode,
-				       struct dentry *dentry,
-				       bool may_sleep)
+static struct inode_security_struct *
+__inode_security_revalidate(struct inode *inode,
+			    struct dentry *dentry,
+			    bool may_sleep)
 {
-	struct inode_security_struct *isec = inode_security_novalidate(inode);
+	struct inode_security_struct *isec = find_isec_ns(inode, may_sleep);
 
 	might_sleep_if(may_sleep);
 
 	if (selinux_initialized(&selinux_state) &&
 	    isec->initialized != LABEL_INITIALIZED) {
 		if (!may_sleep)
-			return -ECHILD;
+			return ERR_PTR(-ECHILD);
 
 		/*
 		 * Try reloading the inode security label.  This will fail if
@@ -372,17 +422,12 @@ static int __inode_security_revalidate(struct inode *inode,
 		 */
 		inode_doinit_with_dentry(inode, dentry);
 	}
-	return 0;
+	return isec;
 }
 
 static struct inode_security_struct *inode_security_rcu(struct inode *inode, bool rcu)
 {
-	int error;
-
-	error = __inode_security_revalidate(inode, NULL, !rcu);
-	if (error)
-		return ERR_PTR(error);
-	return inode_security_novalidate(inode);
+	return __inode_security_revalidate(inode, NULL, !rcu);
 }
 
 /*
@@ -390,8 +435,7 @@ static struct inode_security_struct *inode_security_rcu(struct inode *inode, boo
  */
 static struct inode_security_struct *inode_security(struct inode *inode)
 {
-	__inode_security_revalidate(inode, NULL, true);
-	return inode_security_novalidate(inode);
+	return __inode_security_revalidate(inode, NULL, true);
 }
 
 static struct inode_security_struct *backing_inode_security_novalidate(struct dentry *dentry)
@@ -408,8 +452,7 @@ static struct inode_security_struct *backing_inode_security(struct dentry *dentr
 {
 	struct inode *inode = d_backing_inode(dentry);
 
-	__inode_security_revalidate(inode, dentry, true);
-	return inode_security_novalidate(inode);
+	return __inode_security_revalidate(inode, dentry, true);
 }
 
 static void isech_free_rcu(struct rcu_head *head)
@@ -2967,7 +3010,7 @@ static int selinux_inode_alloc_security(struct inode *inode)
 	struct inode_security_struct *isec;
 	struct inode_security_head *isech = selinux_inode(inode);
 
-	isec = isec_alloc(inode);
+	isec = isec_alloc(inode, true);
 	if (!isec)
 		return -ENOMEM;
 
-- 
2.17.1

